extensions/libraries/joomla_entity/src/Fields/Traits/HasFields.php

```diff
@@ -79,50 +97,92 @@ public function fieldValue($id, $default = null, $raw = false)
 			throw new \InvalidArgumentException($msg);
 		}
 
-		return (null === $values[$id]) ? $default : $values[$id];
+		return (null === $values[$id]) ? $default : $values[$id][$raw ? 'rawvalue' : 'value'];
 	}
 
 	/**
-	 * Get all the field values for this entity.
+	 * Retrieve a field value by a given field name
 	 *
-	 * @param   boolean  $raw  Return raw field values
+	 * @param   integer  $name     Name of the field which value we want to retrieve.
+	 * @param   mixed    $default  Value to use as default if value is null
+	 * @param   boolean  $raw      Return raw field value
+	 *
+	 * @return  mixed
+	 *
+	 * @throws  \InvalidArgumentException  If field value is not set
+	 */
+	public function fieldValueByName($name, $default = null, $raw = false)
+	{
+		$values = $this->fieldValues();
+		$id = array_search($name, array_column($values, 'name'));
+
+		if ($id === false)
+		{
+			return $default;
+		}
+
+		return array_values($values)[$id][$raw ? 'rawvalue' : 'value'];
+	}
+
+	/**
+	 * Get all the field values for this entity.
 	 *
 	 * @return  array
 	 */
-	public function fieldValues($raw = false)
+	public function fieldValues()
 	{
-		$values = array();
-		$fields = $this->fields();
+		// Returns the cached array of values
+		if (null !== $this->fieldValues)
+		{
+			return $this->fieldValues;
+		}
+
+		$fields = $this->fields(false, true);
 
 		if ($fields->isEmpty())
 		{
-			return $values;
+			$this->fieldValues = array();
+
+			return $this->fieldValues;
 		}
 
-		foreach ($fields as $field)
+		$this->fieldValues = array();
+
+		/** @var Field $field */
+		$field = $fields->rewind();
+
+		while ($field)
 		{
-			$property = $raw ? 'rawvalue' : 'value';
-			$values[$field->id()] = $field->get($property);
+			$this->fieldValues[$field->id] = array(
```

Comment:
This is breaking backward compatibility because this method now expects a parameter to be passed and the direct response to be returned.

I like the idea of having a `loadFieldValues()` loading all the data and then maybe have two methods: 

* `fieldValues()` working as it's working now without breaking it
* `fieldRawValues()` returning RAW data. 

That way we can get rid of $raw everywhere without breaking anything.

Reply:
Actually we could use fields for that which would be perfect. Something like:

* $field->getValue()
* $field->getRawValue()

So once fields are initialised we can relay in them. I also like that because a custom entity that represents a field could add logic there to prepare values, etc.